content/curried-varargs.md

```diff
@@ -0,0 +1,271 @@
+
+## Motivating Examples
+
+### Examples
+
+Recently I was working on the implementation of [Pre SIP: name based XML literals](https://contributors.scala-lang.org/t/pre-sip-name-based-xml-literals/2175). During implementing that proposal, I found that the proposal is inefficiency due to repeated parameters, and it could be improved dramatically with the help of curried functions.
+
+For example, according to the proposal the XML literal `<div title="my-title">line1<br/>line2</div>` will result the following code:
+
+``` scala
+xml.tags.div(
+  xml.attributes.title(xml.text("my-title")),
+  xml.text("line1"),
+  xml.tags.br(),
+  xml.text("line2")
+)
+```
+
+With the help of this curried varargs proposal and `@inline`, we are able to implement an API to build a DOM tree with no additional overhead over manually written Scala code.
+
+``` scala
+import org.scalajs.dom.document
+import org.scalajs.dom.raw._
+
+object xml {
+  type Attribute[-A <: Element] = A => Unit
+  @inline def text(data: String) = data
+  object attributes {
+    @inline def title(value: String): Attribute[Element] = _.setAttribute("title", value)
+  }
+  object tags {
+    class Builder[+E <: Element](private val element: E) extends AnyVal with Curried {
+      @inline def applyBegin = this
+      @inline def applyNext(text: String) = {
+        element.appendChild(document.createTextNode(text))
+        this
+      }
+      @inline def applyNext(node: Node) = {
+        element.appendChild(node)
+        this
+      }
+      @inline def applyNext[A <: Attribute[E]](attribute: A) = {
+        attribute(element)
+        this
+      }
+      @inline def applyEnd = element
+    }
+    @inline def div = new Builder(document.createElement("div"))
+    @inline def br = new Builder(document.createElement("br"))
+  }
+}
+```
+
+Since `xml.tags.div` returns a `Builder`, which is a subtype of `Curried`, calls on `xml.tags.div` will be translated to the curried form, as shown below:
+``` scala
+xml.tags.div
+  .applyBegin
+  .applyNext(xml.attributes.title(xml.text("my-title")))
+  .applyNext(xml.text("line1"))
+  .applyNext(xml.tags.br.applyBegin.applyEnd)
+  .applyNext(xml.text("line2"))
+  .applyEnd
+```
+
+When the above code is compiled in Scala.js, the builders should be eliminated entirely as a zero cost abstraction layer, and the output JavaScript is tiny as shown below:
+
+``` javascript
+var $$this = $m_Lorg_scalajs_dom_package$().document__Lorg_scalajs_dom_raw_HTMLDocument().createElement("div");
+$$this.setAttribute("title", "my-title");
+$$this.appendChild($m_Lorg_scalajs_dom_package$().document__Lorg_scalajs_dom_raw_HTMLDocument().createTextNode("line1"));
+var $$this$1 = $m_Lorg_scalajs_dom_package$().document__Lorg_scalajs_dom_raw_HTMLDocument().createElement("br");
+$$this.appendChild($$this$1);
+$$this.appendChild($m_Lorg_scalajs_dom_package$().document__Lorg_scalajs_dom_raw_HTMLDocument().createTextNode("line2"));
```

Comment:
This level of optimization works because Scala.js has a whole-program optimizer that very capable. Do we achieve the same level of optimization on Scala/JVM? Do we have any evidence that this rewrite is not JS-specific?